### DISCUSSION TITLE
Allow dynamic arguments to be used on local functions with inferred generic parameters

### DISCUSSION THREAD

**Discussion by mikernet** [dotnet/csharplang#2051] [Language Ideas]
Is the limitation against using `dynamic` on local generic functions with inferred type parameters due to a large technical hurdle or just a matter of nobody identifying compelling use cases for it yet?

Being able to do that would let me keep things better contained in situations like this, which I'm running into fairly often right now to get around [generic inference limitations](https://stackoverflow.com/questions/53683564/fluent-interface-for-generic-type-hierarchy) and [lack of generic wildcards](https://github.com/dotnet/csharplang/issues/1992):

```c#
    public static TField Required<TField>(this TField field) where TField : Field
    {
        AddRequiredValidator((dynamic)field);
        return field;
    }

    private static void AddRequiredValidator<T>(Field<T> field)
    {
        field.AddValidator(new RequiredValidator<T>());
    }
```

I want to be able to do this:

```c#
    public static TField Required<TField>(this TField field) where TField : Field
    {
        void AddTo<T>(Field<T> typedField) => typedField.AddValidator(new RequiredValidator<T>());
        AddTo((dynamic)field);
        return field;
    }
```

This is a simple example but there are more complicated flows where this would really help keep things better organized and more linear.

**Comment by HaloFour** [comment]
What limitation?  I can use `dynamic` with local functions just fine.  You're not defining your local function to be generic, so I'm not sure why you expect it to behave like the generic instance method in your first example.  It's kind of hard to tell what you're trying to do, though.

**Comment by svick** [comment]
@HaloFour I think the code above has a typo and the goal here is to be able to write:

```c#
public static TField Required<TField>(this TField field) where TField : Field
{
    void AddTo<T>(Field<T> typedField) => typedField.AddValidator(new RequiredValidator<T>());
    AddTo((dynamic)field);
    return field;
}
```

[This currently fails with:](https://sharplab.io/#v2:D4AQjAbABCBMUGECwAoA3qqWYGYaSgBUAxASwFMAbAEygCVyBHAV1ICdzqAeEimgPgAUAFwAWpAM5EyVWgDM+1AJRQA7qPIdpiqAC4oMmpmwYU2czAAsUAILVqhAPY8hh7oX5RhATwAOnNxUAXk8ff2o3ADo7agA1AENKUmp44Uc2QQA7clV6JlYOOMTk1PSXQSUlAG5jC1t7J0FBam9M+IBbUgBjJQVZatqLEAB2KD6aGrNsAF9UWZRUOAMdNHnF+DcXPWXZVFMASBA8EGsYhKSUtIyGFnZOc5KrrYB9FVW51HW828KHy7KPFBVkA==)

> error CS8322: Cannot pass argument with dynamic type to generic local function 'AddTo' with inferred type arguments.

**Comment by mikernet** [comment]
@HaloFour Yeah sorry @svick caught the typo. I'm just trying to turn the second function into a local function inside the first function.

**Comment by mikernet** [comment]
I didn't realize the limitation was only applicable to local **generic** functions...fixed now.

**Comment by HaloFour** [comment]
Ah, I see.  Given the very specific error message it seems quite intentional.  I'd make the assumption that the team was aware that such inference was legal with non-local functions.  Maybe they consider that to be a mistake?

**Comment by mikernet** [comment]
@HaloFour  I think the issue is more that calls to local functions with dynamic parameters don't emit any actual dynamic code but rather a static call that casts the dynamic parameters. I'm guessing it would have been extra work to make sure the proper dynamic code was emitted in the case of a generic method with inferred arguments. Local functions can't have overloads so converting it to a static call works fine as long as there aren't any inferred generic parameters.
